pkg/apis/pipeline/v1beta1/pipeline_validation_test.go

```diff
@@ -1039,10 +1039,395 @@ func TestValidatePipelineWorkspaces_Failure(t *testing.T) {
+				Tasks: []PipelineTask{{
+					Name:    "non-final-task",
+					TaskRef: &TaskRef{Name: "bar-task"},
+					Resources: &PipelineTaskResources{
+						Inputs: []PipelineTaskInputResource{{
+							Name: "some-workspace", Resource: "great-resource",
+						}},
+						Outputs: []PipelineTaskOutputResource{{
+							Name: "some-image", Resource: "wonderful-resource",
+						}},
+					},
+					Conditions: []PipelineTaskCondition{{
+						ConditionRef: "some-condition",
+						Resources: []PipelineTaskInputResource{{
+							Name: "some-workspace", Resource: "great-resource",
+						}},
+					}},
+				}},
+				Finally: []PipelineTask{{
+					Name:    "foo",
+					TaskRef: &TaskRef{Name: "foo-task"},
+					Resources: &PipelineTaskResources{
+						Inputs: []PipelineTaskInputResource{{
+							Name: "some-workspace", Resource: "great-resource",
+						}},
+						Outputs: []PipelineTaskOutputResource{{
+							Name: "some-image", Resource: "wonderful-resource",
+						}},
+					},
+				}},
+			},
+		},
+		failureExpected: false,
+	}, {
+		name: "invalid pipeline without any non-final task but at least one final task",
+		p: &Pipeline{
+			ObjectMeta: metav1.ObjectMeta{Name: "pipeline"},
+			Spec: PipelineSpec{
+				Tasks: nil,
+				Finally: []PipelineTask{{
+					Name:    "final-task",
+					TaskRef: &TaskRef{Name: "final-task"},
+				}},
+			},
+		},
+		failureExpected: true,
+	}, {
+		name: "invalid pipeline without any non-final task but at least one final task",
+		p: &Pipeline{
+			ObjectMeta: metav1.ObjectMeta{Name: "pipeline"},
+			Spec: PipelineSpec{
+				Tasks: []PipelineTask{{}},
+				Finally: []PipelineTask{{
+					Name:    "final-task",
+					TaskRef: &TaskRef{Name: "final-task"},
+				}},
+			},
+		},
+		failureExpected: true,
+	}, {
+		name: "invalid pipeline with empty finally section",
+		p: &Pipeline{
+			ObjectMeta: metav1.ObjectMeta{Name: "pipeline"},
+			Spec: PipelineSpec{
+				Tasks: []PipelineTask{{
+					Name:    "non-final-task",
+					TaskRef: &TaskRef{Name: "non-final-task"},
+				}},
+				Finally: []PipelineTask{{}},
+			},
+		},
+		failureExpected: true,
+	}, {
+		name: "invalid pipeline with duplicate final tasks",
+		p: &Pipeline{
+			ObjectMeta: metav1.ObjectMeta{Name: "pipeline"},
+			Spec: PipelineSpec{
+				Tasks: []PipelineTask{{
+					Name:    "non-final-task",
+					TaskRef: &TaskRef{Name: "non-final-task"},
+				}},
+				Finally: []PipelineTask{{
+					Name:    "final-task",
+					TaskRef: &TaskRef{Name: "final-task"},
+				}, {
+					Name:    "final-task",
+					TaskRef: &TaskRef{Name: "final-task"},
+				}},
+			},
+		},
+		failureExpected: true,
+	}, {
+		name: "invalid pipeline with same task label for final and non final task",
+		p: &Pipeline{
+			ObjectMeta: metav1.ObjectMeta{Name: "pipeline"},
+			Spec: PipelineSpec{
+				Tasks: []PipelineTask{{
+					Name:    "same-task-label",
+					TaskRef: &TaskRef{Name: "non-final-task"},
+				}},
+				Finally: []PipelineTask{{
+					Name:    "same-task-label",
```

Comment:
note that "label" means something specific in k8s resources, might make more sense to call this and the test case "name"

Reply:
thanks @bobcatfish for pointing it out, Done.